table/server/common/src/main/java/alluxio/table/common/udb/UdbMountSpec.java

```diff
@@ -0,0 +1,589 @@
+    SimpleNamePatternWrapper(Set<String> names, Set<Pattern> patterns) {
+      super(names, patterns);
+    }
+
+    static SimpleNamePatternWrapper empty() {
+      return new SimpleNamePatternWrapper(Collections.emptySet(), Collections.emptySet());
+    }
+  }
+
+  abstract static class NamePatternWrapper {
+    protected final ImmutableSet<String> mNames;
+    protected final ImmutableSet<Pattern> mPatterns;
+
+    NamePatternWrapper(Set<String> names, Set<Pattern> patterns) {
+      mNames = ImmutableSet.copyOf(names);
+      mPatterns = ImmutableSet.copyOf(patterns);
+    }
+
+    boolean isEmpty() {
+      return mNames.size() == 0 && mPatterns.size() == 0;
+    }
+
+    boolean has(String name) {
+      return hasExplicit(name) || hasCoveredByPattern(name);
+    }
+
+    boolean hasExplicit(String name) {
+      return mNames.contains(name);
+    }
+
+    boolean hasCoveredByPattern(String name) {
+      return mPatterns.stream().map(p -> p.matcher(name)).anyMatch(Matcher::matches);
+    }
+
+    Set<String> getNames() {
+      return mNames;
+    }
+
+    Set<Pattern> getPatterns() {
+      return mPatterns;
+    }
+  }
+
+  /**
+   * Base builder for common builder methods.
+   * @param <T> the output type
+   * @param <SELFT> the subtype of deriving class
+   */
+  public abstract static class BaseBuilder<T, SELFT extends BaseBuilder<T, SELFT>> {
+    @Nullable
+    protected Set<String> mNames;
+    @Nullable
+    protected Set<Pattern> mPatterns;
+
+    protected BaseBuilder() {}
+
+    /**
+     * Returns this instance.
+     * @return this instance
+     */
+    abstract SELFT self();
+
+    /**
+     * Adds a name.
+     * @param name name
+     * @return the builder
+     */
+    public SELFT addName(String name) {
+      mNames = addOrNewSet(mNames, name);
+      return self();
+    }
+
+    /**
+     * Adds a pattern.
+     * @param pattern pattern
+     * @return the builder
+     */
+    public SELFT addPattern(Pattern pattern) {
+      mPatterns = addOrNewSet(mPatterns, pattern);
+      return self();
+    }
+
+    /**
+     * Adds a set of names, the set has to be mutable.
+     * @param names set of names
+     * @return the builder
+     */
+    public SELFT addNames(Set<String> names) {
+      mNames = addOrNewSet(mNames, names);
+      return self();
+    }
+
+    /**
+     * Adds a set of patterns, the set has to be mutable.
+     * @param patterns set of names
+     * @return the builder
+     */
+    public SELFT addPatterns(Set<Pattern> patterns) {
+      mPatterns = addOrNewSet(mPatterns, patterns);
+      return self();
+    }
+
+    /**
+     * Builds the type.
+     * @return the built type
+     */
+    abstract T build();
+
+    private static <T> Set<T> addOrNewSet(Set<T> someSet, T element) {
+      if (someSet == null) {
+        Set<T> newSet = new HashSet<>();
+        newSet.add(element);
+        return newSet;
+      } else {
+        someSet.add(element);
```

Comment:
Caveat: we cannot know for sure if the passed-in set is mutable or not. This might cause some unsupported operation exception.